consensus/snapshot/snapshots_service.c

```diff
@@ -0,0 +1,420 @@
+
+  log_info(logger_id, "Persisting snapshot with index: %" PRIu64 "\n", snapshot->metadata.index);
+  ERR_BIND_GOTO(iota_snapshots_provider_write_snapshot_to_file(
+                    snapshot, snapshots_service->conf->local_snapshots.local_snapshots_path_base),
+                ret, cleanup);
+
+  log_info(logger_id, "Local snapshot files were succesfully written\n");
+  iota_snapshot_write_lock(&snapshots_service->snapshots_provider->inital_snapshot);
+
+  ERR_BIND_GOTO(iota_snapshot_copy(snapshot, &snapshots_service->snapshots_provider->inital_snapshot), ret, cleanup);
+  log_info(logger_id, "Local snapshot updated with snapshot index %" PRIu64 "\n",
+           snapshots_service->snapshots_provider->inital_snapshot.metadata.index);
+
+cleanup:
+  iota_snapshot_unlock(&snapshots_service->snapshots_provider->inital_snapshot);
+  return ret;
+}
+
+static retcode_t check_transaction_is_not_orphan_do_func(flex_trit_t *hash, iota_stor_pack_t *pack, void *data,
+                                                         bool *should_branch, bool *should_stop) {
+  *should_stop = false;
+  *should_branch = true;
+  check_not_orphan_do_func_params_t *params = data;
+
+  if (pack->num_loaded == 0) {
+    return RC_OK;
+  }
+
+  if (hash_to_uint64_t_map_contains(params->solid_entry_points, hash)) {
+    params->is_orphan = false;
+    *should_stop = true;
+    return RC_OK;
+  }
+
+  uint64_t current_arrival_timestamp = transaction_arrival_timestamp((iota_transaction_t *)pack->models[0]);
+  uint64_t current_timestamp = transaction_timestamp((iota_transaction_t *)pack->models[0]);
+
+  if (current_arrival_timestamp > params->target_milestone_timestamp * 1000UL ||
+      current_timestamp > params->target_milestone_timestamp) {
+    params->is_orphan = false;
+    *should_stop = true;
+  }
+
+  return RC_OK;
+}
+
+static retcode_t find_solid_entry_points_and_update_do_func(flex_trit_t *hash, iota_stor_pack_t *pack, void *data,
+                                                            bool *should_branch, bool *should_stop) {
+  retcode_t ret = RC_OK;
+  *should_stop = false;
+  *should_branch = false;
+  find_solid_entry_points_and_update_do_func_params_t *params = data;
+
+  if (pack->num_loaded == 0) {
+    return RC_OK;
+  }
+
+  uint64_t current_snapshot_index = transaction_snapshot_index((iota_transaction_t *)pack->models[0]);
+
+  // TODO - remove
+  ERR_BIND_RETURN(hash_to_uint64_t_map_add(params->solid_entry_points, hash, current_snapshot_index), ret);
+  return RC_OK;
+
+  // Transaction was referenced by a newer milestone than "target_milestone", therefor, this is a relevant
+  // entry point!
+  if (current_snapshot_index > params->target_milestone_index) {
+    ERR_BIND_RETURN(hash_to_uint64_t_map_add(params->solid_entry_points, hash, current_snapshot_index), ret);
+  }
+
+  if (current_snapshot_index >= params->min_snapshot_index) {
+    *should_branch = true;
+    ERR_BIND_GOTO(iota_snapshots_service_add_entry_point_if_not_orphan(
```

Comment:
So we end up doing a DFS to future while in the DFS to past ? 💥

Reply:
yes